secrecy/src/lib.rs

```diff
@@ -345,4 +345,60 @@ mod tests {
         let secret = SecretString::from_str("test").unwrap();
         assert_eq!(secret.expose_secret(), "test");
     }
+
+    #[cfg(feature = "test-allocator")]
+    mod secret_with_custom_allocator {
+
+        use super::super::*;
+        use core::mem::size_of;
+        use core::ptr;
+
+        use std::alloc::{GlobalAlloc, Layout, System};
+        // Allocator that leaks all memory it allocates, thus leaving the memory open for inspection.
+        struct UnfreeAllocator;
+        #[allow(unsafe_code)]
+        unsafe impl GlobalAlloc for UnfreeAllocator {
+            unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+                System.alloc(layout)
+            }
+            unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+                let _ = (ptr, layout);
+            }
+        }
+
+        #[global_allocator]
+        static UNFREE_ALLOCATOR: UnfreeAllocator = UnfreeAllocator;
+
+        #[test]
+        #[allow(unsafe_code, unused_assignments)]
+        fn clears_memory_when_scope_ends() {
+            let mut ptr: *const u128 = ptr::null();
+            unsafe {
+                {
+                    let secret = SecretBox::init_with(|| 0xdeadbeef_u128);
+                    let secretboxptr = &secret as *const SecretBox<u128>;
+                    // Points to the inner_secret of the `SecretBox`
+                    let boxptr = secretboxptr as *const *const u128;
+                    // Pointer to actual heap data
+                    ptr = *boxptr;
+
+                    assert!(!ptr.is_null(), "ptr is null before drop, not ok");
+                    let bytes: &[u8] =
+                        core::slice::from_raw_parts(ptr as *const u8, size_of::<u128>());
+                    assert!(
+                        !bytes.iter().all(|&b| b == 0),
+                        "Expected non-zero data, instead found 0s: {:X?}",
+                        bytes
+                    );
+                }
+                // Check that the memory is cleared after the scope ends
+                let bytes: &[u8] = core::slice::from_raw_parts(ptr as *const u8, size_of::<u128>());
+                assert!(
+                    bytes.iter().all(|&b| b == 0),
```

Comment:
This is UB since you're making a Rust reference that's UAF.

You could avoid UB by performing only raw pointer operations, i.e. computing new pointers and dereferencing them.